#エラーハンドリングの歴史

## <a name="introduction" href="#introduction">はじめに</a>

2011年現在、ソフトウェアの複雑化により、プログラムは正常系の動作だけができればいい時代は終わっている。

###ゲームにおけるエラーハンドリング
スタンドアロンで動作するゲームが主流であったころ、ゲーム開発者はパフォーマンスのためにライブラリを含む全てのプログラムからエラー処理を省き、デバッグ時にassertで全てのバグを取り去り、正常系のみのプログラムをリリースしていた。現代のゲームは通信処理や、OS・コントローラからのイベントなどにより、正常系のみでは対処できずエラーハンドリングを正しく行う必要が出てきた。

###ソフトウェアにおけるエラーハンドリング
ソフトウェアは、元々マルチタスクやメモリの制限などにより、昔からエラーハンドリングは行われてきたが、
現代のソフトウェアはユーザー人口がとても多かったり、クレジットカードや個人情報の扱いなどで機密性がより求められるようになってきたため、
エラーハンドリングの重要性は一層高くなってきた。


##エラーハンドリングは軽視されがち

エラーハンドリングは、多くの理由から軽視されがちだ。
たとえば以下のような理由が考えられる：

- 正常系のプログラムを書くだけで手一杯で異常系を書く期間がとれない
- 予算の削りどころ(セキュリティやテストなども削りどころ)
- 想定不足


##エラーハンドリングの歴史

エラーハンドリングがどのように進化してきたかを見ていこう。


### <a name="sucess-value-error-value" href="#sucess-value-error-value">1. 正常値とエラー値</a>

いつからかは把握していないが、昔は各ライブラリがそれぞれ異なる規定の下、「0を正常値と見なす」はたまた「1を正常値と見なす」「それ以外の値をエラー値とする」といった決まりごとを作った。
たとえば、Xライブラリ(仮)は「正常値であれば0、それ以外の値はエラーである」という仕様にした。

```cpp
int f(int x)
{
    if (x == 1)
        return 0;
    else
        return 1;
}
```

ユーザーはその決まりごとをドキュメントで確認し、気をつけてコーディングする必要があった。

```cpp
int result = f(3);
if (result == 0) {
    // 正常系の処理
}
else {
    // 異常系の処理
}
```

この方法は、いろいろなライブラリを混ぜて使う場合に問題が起こりやすくなる。
Yライブラリ(仮)が「正常値であれば1、マイナス値はエラーである」という仕様にしたとしよう。

```cpp
int g(x == 1)
{
    if (x == 1)
        return 1;
    else
        return -65536;
}
```

このような、規定の異なる複数のライブラリを同時に使う場合、もしくは規定の異なる別のライブラリを使ったことがあるという経験が、
正常値と異常値の判定ミスというヒューマンエラーの原因になりえた。


### <a name="null-pointer" href="#null-pointer">2. ヌルポインタ</a>

主にC言語では、正常値として有効な値を指すポインタ、異常値としてヌルポインタを渡す／返す設計がよく行われていた。
ここでの問題は、C言語におけるプログラミングのイディオムとして、大きいデータのコピーによるオーバーヘッドを避けるために
データをなんでもかんでもポインタで渡していたことだ。
これのせいで、「この関数のパラメータはヌルポインタでありえていいのか」が非常にわかりにくかった。

また、ヌルポインタには「どこでどんな理由でヌルポインタになったのか」という情報がないため、エラーが追いにくい傾向がある。


### <a name="assert" href="assert">3. アサート</a>
アサートは、「この関数のパラメータがヌルポインタになることはあってはならない」というような表明を行うための機能だ。
これはデバッグ時に取り除くべきエラーを検知するために使用する。
この機能はリリース時には無効となるため、リリース時にも起こりえるエラーなのかどうかの設計判断が重要となる。

```cpp
int f(int* p)
{
    assert(p != NULL);
}
```


### <a name="enumerate" href="#enumerate">4. 列挙値</a>

これまでは、エラーとなる理由を表現するためにエラーコードという整数値が使われてきた。
Windows APIの`GetLastError()`関数がその代表的な例だ。

- [GetLastError() メッセージ一覧](http://homepage3.nifty.com/aokura/tips/e000000.html)

しかし、単なる整数値では、ユーザーはまったく関係のない値との比較を書いてしまう可能性がある。

```cpp
int error_code = GetError();

switch (error_code) {
    case 0: return "正常";
    case 1: return "接続失敗！";
    case 2: return "送信失敗！";
    case 3: return "受信失敗！";

    default: // 何が入ってくるかわからない
        return error_code + ": 不明なエラー";
}
```

そこで、列挙値というものが登場し、現れる可能性を列挙して、その範囲の値しか入り得ないようにすることができるようになった。

```cpp
enum ErrorValue {
    Success,
    ConnectError,
    SendError,
    ReceiveError
};

ErrorValue error_code = GetError();

switch (error_code) {
    case Success:      return "正常";
    case ConnectError: return "接続失敗！";
    case SendError:    return "送信失敗！";
    case ReceiveError: return "受信失敗！";
}
```

これによって、想定できる値をすべて挙げることができる場合において、
意図しない値が代入される可能性をなくすことができた。


### <a name="boolean" href="#boolean">5. 論理値</a>
`bool`(C++), `boolean`(Java), `Bool`(Haskell)のように、いろんな言語で採用されている論理値と呼ばれるものは、正常値を表す値(`true`)と、異常値を表す値(`false`)の2つのみがありえる型とその値である。
これは、現代においてあらゆる環境でデフォルトで使用されるエラーハンドリングとなっている。

```cpp
bool b1 = true;  // 正常値
bool b2 = false; // 異常値
```

論理値の問題点としては、異常値が「なぜ異常になったのか」という情報が持てないということだ。


### <a name="exception" href="exception">6. 例外</a>
例外は、深い階層で発生したエラーを上位層でハンドリングするために、エラー値を伝搬していくのが手間であるという動機の元に導入された。

```cpp
bool f()
{
    return false;
}

bool g()
{
    ...
    bool result = f(); // 関数g()は、関数f()の結果を返す
    ...
    return result;
}

void foo()
{
    bool result = g();
}
```

このプログラムでは、深い階層のどこでエラーが発生したかわかりにくく、また関数`g()`の直接のエラーでないにも関わらず、関数`g()`は関数`f()`の結果を返す必要がある。
これは例外を使うと以下のように書ける：

```cpp
void f()
{
    throw runtime_error("fでエラーがでた");
}

void g()
{
    ...
    f();
    ...
}

void foo()
{
    // エラーが起こりえるブロック
    try {
        g();
    }
    // tryブロック内で例外が投げられたら捕捉する
    catch (runtime_error& e) {
        error_message(e.what()); // 例外メッセージを表示する
    }
}
```

例外は、いくつかのケースでプログラムをわかりやすくした。
C++では例外におけるオーバーヘッドが大きいため、サイズ制限などが厳しい環境では例外が使えない場合が多い。

例外の問題点としては、捕捉ミスによるアプリケーションエラーが起こりやすいということと、どの関数がどの例外を投げうるのかがわかりにくいということだ。
Javaでは検査例外という機構を用意し、「この関数はこの例外を投げうる」ということを明示することができたが、近年のジェネリックプログラミングの流れにより、「この関数に渡されるあらゆる型の同名の関数が全てこの例外を投げうる」というのを明示するのは難しいと判断され、ScalaのようなJavaの後継言語では検査例外は採用されなかった。


参照：

- [例外設計における大罪](http://www.slideshare.net/t_wada/exception-design-by-contract)
- [例外設計の話。](https://gist.github.com/sunaot/6138546#file-exception-md)


### <a name="option" href="option">7. 無効値の特別な表現</a>
C++, Haskell, Scalaのような静的型付け言語では、エラーが起こりえる型であることを明示する方向に進んだ。
ここでは「[8. パターンマッチ](#pattern-match)」との差別化のため、C++における`boost::optional<T>`型で説明する。

「[1. 正常値とエラー値](#success-value-error-value)」で解説した「正常値／エラー値の仕様がライブラリによってまちまちで、仕様を調べないとミスが起こりやすい」という問題に対処するため、Boost C++ Librariesでは、`boost::optional<T>`という型を提供することにより、型によらない無効値の表現と、その型の値。この2種類が起こりえる型を表現した。
これにより、正常値、エラー値というものをライブラリの仕様ではなく、エラーが起こりえる型という形で、ヒューマンエラーの軽減に取り組んだ。

```cpp
boost::optional<int> a = 1; // int型の値全てが正常値
boost::optional<int> b = boost::none; // noneが無効値

if (a) { // 正常値かどうかの判定
    int x = a.get(); // 中身を取り出す
}
```

代表的な例：

| 言語    | 表現 |
|---------|------|
| C++     | `boost::optional<T>` |
| Haskell | `Maybe` |
| Scala   | `Option` |
| C#      | null許容型 |


### <a name="pattern-match" href="pattern-match">8. パターンマッチ</a>
パターンマッチは、 「[7. 無効値の特別な表現][#option]」をさらに進化させ、型の判別と分解を同時に行うことができる。
`boost::optional<T>`型は、値が正常値かどうかの判定と、中身を取り出す操作がわかれているため、無効値を指した状態で中身を取り出すとアプリケーションエラーになるだろう。

関数型言語一般で採用されているパターンマッチと呼ばれる機構は、この判定と取り出しを同時に行うことで、ヒューマンエラーを除去することができる。
Scala言語の`Option`型で例示する：

```scala
a : Option[int] = Some(1)
b : Option[int] = None

match (a) {
    case Some(x) => print x
    case _       => print "error"
}
```

ここでは、`case Some(x)`という式によって、「正常値だった場合には中身を取り出して変数xに代入する」ということをしている。
これによって、無効値を指しているのに中身を取り出そうと間違うことができなくなった。


参考：

- [kinaba さんが Cryolite を洗脳してパターンマッチ厨に仕立て上げるリスト](http://togetter.com/li/6990)


### <a name="adt" href="#adt">9. 代数データ型</a>
C++で論理値を表現するための`bool`型は、歴史的経緯もあって、`true`／`false`以外でも任意の`int`値を格納できてしまい、たびたび意図しない値の代入による問題が起こっていた。

Haskellのような言語では、論理値は代数データ型と呼ばれる機能によって定義される。
`Bool`型を以下のように定義した場合、

```haskell
data Bool = True | False
```

`Bool`型は、「`True`」と「`False`」という2つの値以外は、決して格納できない型となる。

```cpp
f :: Bool
f = True
 
g :: Bool
g = False
 
h :: Bool
h = 0     -- エラー！
```

代数データ型は、「`0`という値が入りえない自然数の型」「要素が空になりえないリスト型」などに応用でき、型によって多くの保証を生み出すことに成功している。


参照：

- [GADTs使ってみた - Faith and Brave - C++で遊ぼう](http://faithandbrave.hateblo.jp/entry/20111201/1322718742)


他の言語でも、列挙型(`enum`)によってある程度表現できる。


### <a name="either" href="either">10. Either</a>

「[7. の無効値と有効値の統一的な表現](#option)」により、エラーが起こりえる箇所がより明確となった。
だが、無効値には「無効値となった理由」という情報が欠落しているため、Haskellの`Maybe`モナドのような「失敗するかもしれない計算を連続で行い、失敗したらその後の計算を行わない」というアプローチにおいて、どの段階でどういう原因でエラーとなったのかがわからず、デバッグが困難であるという問題が起き始めた。

参考： [data Maybe — harmful?](http://blog.dbpatterson.com/post/9528836599)

そこで、エラー情報付きの無効値を表現する方法として、`Either`という型が多くの言語でライブラリとして導入されるようになった。
`Either`は、「`Left`か`Right`どちらかにデータが入っている」というエラー以外の目的にも使用できる普遍的な型ではあるが、エラーハンドリングに使うのが一般的である。

有効値が入っていたら「正しい」を意味する`Right`にデータが入り、無効値が入っていたらLeftにエラーとなった理由を入れる、という文化がある。

代表的な例：

| 言語 | 型 |
|------|----|
| Haskell | `Either` |
| C++     | `expected` |


参照：

- [エラー値と正常値を表す汎用的な型：`expected` - Faith and Brave - C++で遊ぼう](http://faithandbrave.hateblo.jp/entry/2014/05/30/153325)


###11. 自動再起動

now studying Erlang...


###例外安全性


##参照
- [言語女子会: undefとnullは両方必要？ - 西尾泰和のはてなダイアリー](http://d.hatena.ne.jp/nishiohirokazu/20120316/1331906887)


